src/charm.py

```diff
@@ -61,366 +34,195 @@
-        )
-        self.framework.observe(
-            self.on["cassandra_peers"].relation_departed,
-            self.on_cassandra_peers_departed,
-        )
         # TODO: We need to query the version from the database itself
         self.provider = CassandraProvider(charm=self, name="database")
         self.framework.observe(self.provider.on.data_changed, self.on_provider_data_changed)
 
-        self.prometheus_consumer = MetricsEndpointProvider(
+        self.prometheus_provider = MetricsEndpointProvider(
             charm=self,
-            name="monitoring",
-            service_event=self.on.cassandra_pebble_ready,
+            relation_name="monitoring",
             jobs=[
                 {
-                    "static_configs": [{"targets": ["*:{}".format(PROMETHEUS_EXPORTER_PORT)]}],
+                    "static_configs": [{"targets": [f"*:{PROMETHEUS_EXPORTER_PORT}"]}],
                 }
             ],
         )
-        self.framework.observe(self.on["monitoring"].relation_joined, self.on_monitoring_joined)
+        self.framework.observe(self.on["monitoring"].relation_created, self.on_monitoring_created)
         self.framework.observe(self.on["monitoring"].relation_broken, self.on_monitoring_broken)
 
-        self.dashboard_consumer = GrafanaDashboardConsumer(
-            charm=self,
-            name="grafana-dashboard",
-        )
-        self.framework.observe(
-            self.on["grafana-dashboard"].relation_joined, self.on_dashboard_joined
-        )
-        self.framework.observe(
-            self.on["grafana-dashboard"].relation_broken, self.on_dashboard_broken
-        )
-        self.framework.observe(
-            self.dashboard_consumer.on.dashboard_status_changed,
-            self.on_dashboard_status_changed,
-        )
-
-    @status_catcher
-    def on_pebble_ready(self, event):
-        self._configure()
-        container = event.workload
-        if len(self.model.relations["monitoring"]) > 0:
-            self._setup_monitoring()
-        make_started(container)
-        self.provider.update_address("database", self._bind_address())
-
-    @status_catcher
-    def on_config_changed(self, event):
-        self._configure()
-        self.provider.update_port("database", self.model.config["port"])
+        self.dashboard_provider = GrafanaDashboardProvider(self)
+        self.cassandra = Cassandra(charm=self)
+        logging.getLogger("cassandra").setLevel(logging.CRITICAL)
 
-    def on_leader_elected(self, event):
-        self.provider.update_address("database", self._bind_address())
+    def on_pebble_ready(self, event: EventBase) -> None:
+        """Run the pebble ready hook.
 
-    def on_database_joined(self, event):
-        self.provider.update_port("database", self.model.config["port"])
-        self.provider.update_address("database", self._bind_address())
+        Args:
+            event: the event object
+        """
+        if self._num_units() != self._goal_units():
```

Comment:
```suggestion
        if self._num_units() != self.app.planned_units():
```

feels like that `_goal_units` helper isn't providing a lot of value at this point